php-classes/Slate/CBL/StudentCompetency.php

```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
+                '    ON Demonstration.ID = DemonstrationSkill.DemonstrationID '.
+                ' WHERE DemonstrationSkill.SkillID IN (%s) '.
+                '   AND DemonstrationSkill.TargetLevel = %u '.
+                ' ORDER BY SkillID, DemonstrationDate',
+                [
+                    DemonstrationSkill::$tableName,
+                    Demonstration::$tableName,
+                    $this->StudentID,
+                    join(', ', $this->Competency->getSkillIds()),
+                    $this->Level
+                ]
+            );
+
+            return $this->demonstrationData = $data;
+        } catch (\Exception $e) {
+            throw $e;
+            return [];
+        }
+    }
+
+    private $effectiveDemonstrationsData;
+    public function getEffectiveDemonstrationsData()
+    {
+        if (!empty($this->effectiveDemonstrationsData)) {
+            return $this->effectiveDemonstrationsData;
+        }
+
+        $demonstrationsData = $this->getDemonstrationData();
+
+        foreach ($demonstrationsData as $skillId => &$demonstrationData) {
+            $sort = [];
+            foreach ($demonstrationData as $key => $row) {
+                $sort[$key]  = $row['DemonstratedLevel'];
+            }
+            // sort demonstrations by highest level
+            array_multisort($sort, SORT_DESC, $demonstrationData);
+
+            $Skill = Skill::getByID($skillId);
+            $demonstrationsRequired = $Skill->getDemonstrationsRequiredByLevel($this->Level);
+
+            array_splice($demonstrationData, $demonstrationsRequired);
+        }
+
+        return $this->effectiveDemonstrationsData = $demonstrationsData;
+    }
+
+    private $demonstrationsLogged;
+    public function getDemonstrationsLogged()
+    {
+
+        if (!empty($this->demonstrationsLogged)) {
+            return $this->demonstrationsLogged;
+        }
+
+        $effectiveDemonstrationsData = $this->getEffectiveDemonstrationsData();
+        $count = 0;
+        foreach ($effectiveDemonstrationsData as $skillId => $demonstrationData) {
+            $Skill = Skill::getByID($skillId);
+            $skillCount = 0;
+            foreach ($demonstrationData as $demonstration) {
+                if (empty($demonstration['Override'])) {
+                    if (!empty($demonstration['DemonstratedLevel'])) {
+                        $skillCount++;
+                    }
+                }
+            }
+            // ?
+            $count += min($skillCount, $Skill->getDemonstrationsRequiredByLevel($this->Level));
+        }
+
+        return $this->demonstrationsLogged = $count;
+
+    }
+
+    private $demonstrationsComplete;
+    public function getDemonstrationsComplete()
+    {
+        if (!empty($this->demonstrationsComplete)) {
+            return $this->demonstrationsComplete;
+        }
+
+        $effectiveDemonstrationsData = $this->getEffectiveDemonstrationsData();
+        $count = 0;
```

Comment:
just use `$this->demonstrationsComplete`, initialize it here and return it, we don't need to worry about simultaneous access or anything like that